src/config.c

```diff
@@ -173,54 +163,151 @@ config_read_cmd(const char *ipath, const char *set_command_format)
+      fclose(test);
+      config_read_cmd(full_file, set_command_format, rha);
+     }
+   }
 
-  rha = (config_handle *)malloc(sizeof(config_handle));
-  if (rha == NULL)
+  if ((bp = config_get("blah_bin_directory", rha)) != NULL)
    {
-    fclose(cf);
-    free(path);
+    rha->bin_path = strdup(bp->value);
+   }
+  else
+   {
+    rha->bin_path = (char *)malloc(strlen(install_location)+5);
+    if (rha->bin_path != NULL) sprintf(rha->bin_path,"%s/bin",install_location);
+   }
+  if (rha->bin_path == NULL)
+   {
+    /* Out of memory */
+    config_free(rha);
     return NULL;
    }
 
-  rha->config_path = path;
-  rha->list = NULL;
-  if (install_location != NULL) rha->install_path = strdup(install_location);
-  if (rha->install_path == NULL)
+  if ((bp = config_get("blah_sbin_directory", rha)) != NULL)
+   {
+    rha->sbin_path = strdup(bp->value);
+   }
+  else
+   {
+    rha->sbin_path = (char *)malloc(strlen(install_location)+6);
+    if (rha->sbin_path != NULL) sprintf(rha->sbin_path,"%s/sbin",install_location);
+   }
+  if (rha->sbin_path == NULL)
    {
     /* Out of memory */
-    fclose(cf);
     config_free(rha);
     return NULL;
    }
-  rha->bin_path = NULL; /* These may be filled out of config file contents. */
-  rha->sbin_path = NULL; 
-  rha->libexec_path = NULL;
+
+  if ((bp = config_get("blah_libexec_directory", rha)) != NULL)
+   {
+    rha->libexec_path = strdup(bp->value);
+   }
+  else
+   {
+    /* If libexec has a 'blahp' subdirectory, then the blahp files should
+     * be in there.
+     */
+    rha->libexec_path = (char *)malloc(strlen(install_location)+15);
+    if (rha->libexec_path != NULL)
+     {
+      sprintf(rha->libexec_path,"%s/libexec/blahp",install_location);
+      test_dir = opendir(rha->libexec_path);
+      if (test_dir == NULL)
+       {
+        sprintf(rha->libexec_path,"%s/libexec",install_location);
+       } else {
+        closedir(test_dir);
+       }
+     }
+   }
+  if (rha->libexec_path == NULL)
+   {
+    /* Out of memory */
+    config_free(rha);
+    return NULL;
+   }
+
+  return rha;
+ }
+
+int
+config_read_cmd(const char *path, const char *set_command_format, config_handle *rha)
+ {
+  char *line=NULL,*new_line=NULL;
+  char *cur;
+  char *key_start, *key_end, *val_start, *val_end;
+  int key_len, val_len;
+  char *eq_pos;
+  config_entry *c_tail = NULL;
+  config_entry *found,*new_entry=NULL;
+  char *set_command=NULL;
+  int set_command_size;
+  int line_len = 0;
+  int line_alloc = 0;
+  int c;
+  const int line_alloc_chunk = 128;
+  FILE *cf;
+
+  if (path == NULL)
+   {
+    return FALSE;
+   }
+
+  set_command_size = snprintf(NULL,0,set_command_format,path)+1;
+  set_command = (char *)malloc(set_command_size);
+  if (set_command == NULL)
+   {
+    return FALSE;
+   }
+  snprintf(set_command, set_command_size, set_command_format, path);
+
+  cf = popen(set_command, "r");
+  if (cf == NULL) 
+   {
+    free(set_command);
+    return FALSE;
+   }
+
+  free(set_command);
```

Comment:
Minor, but would this be easier on the eyes?

```suggestion
  free(set_command);
  if (cf == NULL) 
   {
    return FALSE;
   }
```